src/main/java/com/cloudbees/jenkins/support/filter/SensitiveContentFilter.java

```diff
@@ -47,37 +50,44 @@
+    private final transient ThreadLocal<Map<String, String>> replacementsMap = new ThreadLocal<>();
+
     public static SensitiveContentFilter get() {
         return ExtensionList.lookupSingleton(SensitiveContentFilter.class);
     }
 
     @Override
     public @NonNull String filter(@NonNull String input) {
-        ContentMappings mappings = ContentMappings.get();
-        String filtered = input;
-        List<String> searchList = new ArrayList<>();
-        List<String> replacementList = new ArrayList<>();
-
-        for (ContentMapping mapping : mappings) {
-            searchList.add(mapping.getOriginal());
-            replacementList.add(mapping.getReplacement());
-        }
-        if (!searchList.isEmpty()) {
-            filtered = WordReplacer.replaceWordsIgnoreCase(input, searchList.toArray(new String[0]), replacementList.toArray(new String[0]));
-        }
-
-        return filtered;
+        return WordReplacer.replaceWords(input, mappingsPattern.get(), replacementsMap.get());
     }
 
     @Override
     public synchronized void reload() {
-        ContentMappings mappings = ContentMappings.get();
+        final long startTime = System.currentTimeMillis();
+        final Map<String, String> replacementsMap = new HashMap<>();
+        final WordsTrie trie = new WordsTrie();
+        final ContentMappings mappings = ContentMappings.get();
         Set<String> stopWords = mappings.getStopWords();
-        for (NameProvider provider : NameProvider.all()) {
+        NameProvider.all().forEach(provider ->
             provider.names()
-                    .filter(name -> StringUtils.isNotBlank(name) && !stopWords.contains(name.toLowerCase(Locale.ENGLISH)))
-                    .forEach(name -> mappings.getMappingOrCreate(name, original -> ContentMapping.of(original, provider.generateFake())));
-        }
+                .filter(StringUtils::isNotBlank)
+                .forEach(name -> {
+                    String lowerCaseOriginal = name.toLowerCase(Locale.ENGLISH);
+                    // NOTE: We could well create a WordTrie for the stop words and use it as a filter instead of the
+                    // conditional here. Or find a better way to deal with insensitive key mapping in general.
+                    // But the reload is already quite fast anyway. (~1s for 10^4 items with 1 CPU / 2 GB memory
+                    // container)
+                    if(!stopWords.contains(lowerCaseOriginal)) {
+                        ContentMapping mapping = mappings.getMappingOrCreate(name, original -> ContentMapping.of(original, provider.generateFake()));
+                        replacementsMap.put(lowerCaseOriginal, mapping.getReplacement());
+                        trie.add(lowerCaseOriginal);
+                    }
+                }));
+        this.mappingsPattern.set(Pattern.compile("\\b" + trie.getRegex() + "\\b", Pattern.CASE_INSENSITIVE));
+        this.replacementsMap.set(replacementsMap);
```

Comment:
I guess this is somehow dealing with concurrent support bundle generation? Suggests an API mistake somewhere—a context should be passed into methods which require it. Maybe tricky to fix now due to use of the API by other plugins?

Reply:
Correct. To deal with concurrent bundle generation. 

The [ContentFilter#reload](https://github.com/jenkinsci/support-core-plugin/blob/master/src/main/java/com/cloudbees/jenkins/support/filter/ContentFilter.java#L62-L66) should be used prior to using the filters. It's always done before starting the bundle generation. So I think this is fine here ?

However, `ThreadLocal` is is a problem when using the filter over remoting. I first thought that a `ThreadLocal` would be well suited, given the ability to process several bundle generation request concurrently. But maybe we should just actually use an `AtomicReference` here. We don't truly need a thread context..

Reply:
> It's always done before starting the bundle generation.

OK. I guess if it takes no args then it is harmless for it to be called again in the middle of bundle generation from another thread starting another bundle?

Reply:
Tested multiple generation in parallel and it works. Should I maybe create a test for the SensitiveFilter with parallel threads ?